pkg/failover-assistant/assistant.go

```diff
@@ -0,0 +1,363 @@
+			task, shutdown := fa.failoverNodeQueue.Get()
+			if shutdown {
+				log.WithFields(log.Fields{"task": task}).Debug("Stop the Node Failover worker")
+				break
+			}
+			if err := fa.processNodeForFailover(task); err != nil {
+				log.WithFields(log.Fields{"task": task, "error": err.Error()}).Error("Failed to process Node Failover task, retry later")
+				fa.failoverNodeQueue.AddRateLimited(task)
+			} else {
+				log.WithFields(log.Fields{"task": task}).Debug("Completed a Node Failover task.")
+				fa.failoverNodeQueue.Forget(task)
+			}
+			fa.failoverNodeQueue.Done(task)
+
+		}
+	}()
+
+	<-stopCh
+	fa.failoverNodeQueue.Shutdown()
+}
+
+func (fa *failoverAssistant) processNodeForFailover(nodeName string) error {
+	logCtx := log.WithField("node", nodeName)
+	logCtx.Debug("Start to failover the stateful pods for the node")
+	node, err := fa.nodeInformer.Lister().Get(nodeName)
+	if err != nil {
+		logCtx.WithError(err).Error("Failed to get node info")
+		return err
+	}
+	if !fa.shouldFailoverForNode(node) {
+		// double check in case that nodes'status is changed back to normal shortly
+		logCtx.Info("Cancel failover for node")
+		return nil
+	}
+
+	pods, err := fa.podInformer.Informer().GetIndexer().ByIndex(podScheduledNodeNameIndex, nodeName)
+	if err != nil {
+		logCtx.WithError(err).Error("Failed to get pods on the node")
+		return err
+	}
+
+	failedPodList := []string{}
+	for i := range pods {
+		pod, _ := pods[i].(*corev1.Pod)
+		if err := fa.failoverForPod(context.TODO(), pod.Spec.NodeName, pod); err != nil {
+			log.WithFields(log.Fields{"namespace": pod.Namespace, "pod": pod.Name}).WithError(err).Debug("Failed to handle pod failover ")
+			failedPodList = append(failedPodList, fmt.Sprintf("%s/%s", pod.Namespace, pod.Name))
+		} else {
+			log.WithFields(log.Fields{"namespace": pod.Namespace, "pod": pod.Name}).Debug("Failed over the pod successfully")
+		}
+	}
+
+	if len(failedPodList) != 0 {
+		return fmt.Errorf("node failover not completed")
+	}
+
+	return fa.completeNodeFailover(node)
+}
+
+func (fa *failoverAssistant) failoverForPod(ctx context.Context, nodeName string, pod *corev1.Pod) error {
+	logCtx := log.WithFields(log.Fields{"node": nodeName, "namespace": pod.Namespace, "pod": pod.Name})
+
+	volFailoverRequests := []*VolumeFailoverRestRequest{}
+	for _, vol := range pod.Spec.Volumes {
+		if vol.PersistentVolumeClaim == nil {
+			continue
+		}
+		volReqs, err := fa.buildVolumeFailoverRequestForPVC(pod.Namespace, nodeName, vol.PersistentVolumeClaim.ClaimName)
+		if err != nil {
+			logCtx.WithField("pvc", vol.PersistentVolumeClaim.ClaimName).WithError(err).Error("Failed to build volume failover requests for pvc")
+			return err
+		}
+		volFailoverRequests = append(volFailoverRequests, volReqs...)
+	}
+
+	logCtx.Debug("Failover for pod")
+	for i := range volFailoverRequests {
+		if err := fa.failoverForVolume(ctx, volFailoverRequests[i]); err != nil {
+			return err
+		}
+	}
+
+	return fa.deletePod(ctx, pod, 0)
```

Comment:
is it ok to `deletePod` first and then `failoverForVolume`?  what bothers me is that the pod is still running but the VolumeAttachment is removed from this node.

perhaps also to consider is how to prevent pods from being re-scheduled to the current node.

Reply:
when to trigger the failover? It should be the case that the pod runs into the problem but can't be deleted successfully. In this case, it has to failover the volume firstly which will clean up the volumeattachment, and then the pod can be deleted automatically or manually.

As to the rescheduling the pod, it should be managed by the admin but not the failover-assistant.